[iOS] Retrieve device locale in full (specific) format from ObjectiveC APIs

matouskozak opened the pull request:
# Problem Description
On iOS (and other apple mobile platforms), we are taking the system default culture and using it inside .NET as `CultureInfo.CurrentCulture`. 

Currently, the system default culture is truncated into neutral culture format (only language). This behavior was set in https://github.com/dotnet/runtime/pull/104071, where we had an incorrect assumption that other ICU platforms also return just neutral locale name. However, that isn't true because the ICU `uloc_getBaseName` https://github.com/dotnet/runtime/blob/117c4ab7723da207ad56014b625395c252af9b6a/src/native/libs/System.Globalization.Native/pal_locale.c#L232 only strips collation information as described in https://github.com/unicode-org/icu/blob/d9d09db2a7167eea2b5749836ef32cdd45303fdd/icu4c/source/common/unicode/uloc.h#L948.

~The problem of using all Apple's system default locales is that some of the retrieved identifiers are not part of https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs (e.g., en-CZ), causing the culture LCID number to be 0x1000 (`LOCALE_CUSTOM_UNSPECIFIED`). This can cause issues where the default `CultureInfo.CurrentCulture` causes crashes when using certain `CultureInfo` APIs that accept LCID instead of culture name, such as, https://github.com/dotnet/runtime/blob/94c356d55a9f9d902f61708ee01e99d480597427/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs#L993~

~used by e.g. `SqlString`.~

# Proposed Change 
This PR reverts the previous change and preserves the full (specific = language-region) culture locale format ~when the retrieved locale is part of `IcuLocaleData.cs`. Upon retrieving the system locale name, we verify that the culture name is part of the ICU data and if not, we use the parent culture (neutral) instead.~

Previous behavior (.net 9):
| Apple system culture | `CurrentCulture` | IsCustomCulture? |
|----------------|------------------|-----------|
| `en-US`             | `en` | no |
| `en-CZ`             | `en` | no |

New behavior:
| Apple system culture | `CurrentCulture` | IsCustomCulture? |
|----------------|------------------|----------|
| `en-US`             | `en-US` | no |
| `en-CZ`             | `en-CZ` | yes |

---

This PR addresses the following issues:
- https://github.com/dotnet/runtime/issues/108958
- https://github.com/dotnet/runtime/issues/110680
- https://github.com/dotnet/runtime/issues/110704

matouskozak committed:
Fix CurrentCulture to be in specific format

matouskozak commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

vitek-karas commented on src/native/libs/System.Globalization.Native/pal_locale.m:
```diff
@@ -786,21 +786,14 @@ int32_t GlobalizationNative_GetLocalesNative(UChar* value, int32_t length)
     }
 }
 
-static NSString* GetBaseName(NSString *localeIdentifier)
-{
-    NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier:localeIdentifier];
-    NSString *languageCode = [locale objectForKey:NSLocaleLanguageCode];
-    return languageCode;
-}
-
```
Do we know what this was done before?
The PR description should be updated to describe this - what we did before (has that) and why, and what we're changing it to (and why it's OK to change given the previous reasoning).

matouskozak replied:
Indeed, I was trying this PR to see which test suit breaks if we switch to specific name because there are other dependencies such as `SqlStrings` that were previously affected. I'll update the issue description when I finalize the code changes.


Brief history what I found:
- The switch to neutral locale names was done in https://github.com/dotnet/runtime/pull/104071
- In that PR, there was an assumption that our ICU also returns just neutral locale name. However, that isn't true as the ICU `uloc_getBaseName` https://github.com/dotnet/runtime/blob/117c4ab7723da207ad56014b625395c252af9b6a/src/native/libs/System.Globalization.Native/pal_locale.c#L232 only strips collation information as described in https://github.com/unicode-org/icu/blob/d9d09db2a7167eea2b5749836ef32cdd45303fdd/icu4c/source/common/unicode/uloc.h#L948.
- The problem with using Apple's system default locale is that if it's not part of https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Globalization/IcuLocaleData.cs, the culture LCID number is 0x1000 (`LOCALE_CUSTOM_UNSPECIFIED`). While it seems that using LCID number to create new cultures is not recommended and using culture name is preferred, e.g., `SqlString` implementation still relies on using it, thus causing failures on apple mobile devices with system locale that is unrecognized by `IcuLocaleData.cs`.

I'll try to make a middle way out of this, where we use system locale if it's recognized by .NET, otherwise we use parent culture which is neutral.

matouskozak commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

matouskozak commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

tarekgh commented:
I don't think we need to fix this. We encourage users to move away from LCIDs. Any users still using LCID should start migrating to use culture names instead. The change here is trying to work around the issue of using the parent which may benefit using the LCID but it can cause other issues when requesting any region-specific data from current culture. 

This issue is not specific to iOS or mobile platforms. This can occur on Windows for example doing `CultureInfo.CurrentCulture = new CultureInfo("en-AE");`. Whoever is using LCID needs to fix their code and start using culture names or they need to make their code tolerant when getting custom LCID values. 

Last, if `IcuLocaleData.cs` missing any data, we can try to update it as needed.

matouskozak commented:
> Any users still using LCID should start migrating to use culture names instead.

@tarekgh I know that using LCID is not recommended. What I was trying to work around here by using the parent is that any default culture set by .NET would be compatible with all .NET library code. We are currently relying on LCID inside `SqlString` codebase, causing failures if LCID is custom culture.

Note, this PR not only does the fallback to parent culture but also fixes the behavior that we didn't return specific system culture. I can remove the parent culture workaround and disable the failing tests until the affected library code migrates from LCID. However, because we need to backport this fix to .NET 9, we would be risking a breaking change to some customers.

tarekgh commented:
> What I was trying to work around here by using the parent is that any default culture set by .NET would be compatible with all .NET libraries code. We are currently relying on LCID inside SqlString codebase, causing failures if LCID is custom culture.

I think the right thing to do is have SqlString get fixed. Either they migrate to use the culture name or to be tolerant to do the fallback themselves. As I mentioned, this issue is not specific to iOS. What happens if someone sets `en-AE` as a default culture on Windows? They will run into the exact same problem. Also, your fix can cause some different problems. Like someone is using the CurrentCulture name to create a RegionInfo object. If you are using the parent (which most likely be a neutral culture, RegionInfo creation will throw exception. 

> but also fixes the behavior that we didn't return specific system culture.

I am fine with the change in `pal_locale.m` if other reviewers see this is a correct fix.

ivanpovazan commented:
> We encourage users to move away from LCIDs

Is there an official documentation about it?

> We are currently relying on LCID inside SqlString codebase, causing failures if LCID is custom culture

Is `SqlString` codebase the only problem? Could we detect this in some way and produce a build warning, so that internally we can fix all the places relying on this?

tarekgh commented:
> Is there an official documentation about it?

https://learn.microsoft.com/en-us/globalization/locale/other-locale-names#lcid

matouskozak commented:
> > We are currently relying on LCID inside SqlString codebase, causing failures if LCID is custom culture
> 
> Is `SqlString` codebase the only problem? Could we detect this in some way and produce a build warning, so that internally we can fix all the places relying on this?

I took a look and there are a few other places where LCID is used but I'm unsure if they would cause the crash. I would need to run each library test suit locally to verify because the test might not fail on CI as the default `CurrentCulture` on lab device is likely `en-US` (non-custom, not causing the before-mentioned crash).

To trigger build warnings, we would need to mark the APIs [obsolete](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-9.0)/depracated. Not sure if this is something that is planned based on https://learn.microsoft.com/en-us/globalization/locale/other-locale-names#lcid @tarekgh ?

matouskozak committed:
Merge branch 'main' into iOS-ICU-return-full-locale

matouskozak commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

matouskozak commented:
fyi: issue for removal of LCID inside `SqlString` codebase https://github.com/dotnet/runtime/issues/111395.

tarekgh commented:
> To trigger build warnings, we would need to mark the APIs [obsolete](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-9.0)/depracated. Not sure if this is something that is planned based on https://learn.microsoft.com/en-us/globalization/locale/other-locale-names#lcid @tarekgh ?

Yes, we should deprecate the LCID property (mark it as obsolete).

matouskozak commented:
/azp run runtime-ioslikesimulator

matouskozak commented:
/azp run runtime-ioslikesimulator

matouskozak committed:
disable the test for now

matouskozak commented:
/azp run runtime-ioslike,runtime-ioslikesimulator,runtime-maccatalyst

matouskozak commented:
I removed the fallback to neutral culture, thus retrieving the locale in full (specific) locale format. 

I also added a test case to assert this behavior. However, for some reason the lab simulators are reporting only `en` instead of expected `en-US`. I verified locally that on simulators (down to iOS 15.5), the full locale identifier is retrieved. It is not the first time I've seen different behavior on lab simulators in comparison to local runs, so I opened an issue to investigate this further https://github.com/dotnet/runtime/issues/111501.

Finally, I opened an issue to address the removal of LCID dependencies inside `SqlString` codebase issue for removal of LCID inside SqlString codebase https://github.com/dotnet/runtime/issues/111395.

Please take a look and review the changes @tarekgh @ivanpovazan @akoeplinger @rolfbjarne

ivanpovazan commented on src/libraries/System.Runtime/tests/System.Globalization.Tests/CultureInfo/CultureInfoCurrentCulture.cs:
```diff
@@ -42,6 +42,10 @@ public void CurrentCulture_Default_Not_Invariant()
             // culture.
             Assert.NotEqual(CultureInfo.CurrentCulture, CultureInfo.InvariantCulture);
             Assert.NotEqual(CultureInfo.CurrentUICulture, CultureInfo.InvariantCulture);
+            
+            // The current culture taken from default system culture should be specific
+            // TODO: https://github.com/dotnet/runtime/issues/111501
+            //Assert.False(CultureInfo.CurrentCulture.IsNeutralCulture);
```
I would suggest instead of adding this as a comment, to make a separate test method with a `Skip` attribute specifying the reason for exception i.e. GH issue you created.

ivanpovazan reviewed (approved):
LGTM, nice work! 
I left a small comment.

matouskozak committed:
move test to separate method

matouskozak commented:
/azp run runtime-ioslikesimulator

matouskozak commented:
/backport to release/9.0-staging
